liquidator/test/Liquidator.js

```
@@ -430,6 +430,188 @@ contract("Liquidator.js", function(accounts) {
+      assert.equal(liquidationObject.liquidator, liquidatorBot);
+      assert.equal(liquidationObject.state, LiquidationStatesEnum.PRE_DISPUTE);
+      assert.equal(liquidationObject.liquidatedCollateral, toWei("33.3333333333333333"));
+      assert.equal(liquidationObject.tokensOutstanding, toWei("4"));
+
+      // Sponsor2 should be in a liquidation state with the bot as the liquidator. (3/8) = 37.5% of the 100 starting collateral and 3 tokens should be liquidated.
+      liquidationObject = (await emp.getLiquidations(sponsor2))[0];
+      assert.equal(liquidationObject.sponsor, sponsor2);
+      assert.equal(liquidationObject.liquidator, liquidatorBot);
+      assert.equal(liquidationObject.state, LiquidationStatesEnum.PRE_DISPUTE);
+      assert.equal(liquidationObject.liquidatedCollateral, toWei("37.5"));
+      assert.equal(liquidationObject.tokensOutstanding, toWei("3"));
+
+      // Sponsor3 should not have been liquidated.
+
+      // Sponsor1 should have some collateral and tokens left in their position from the liquidation.
+      assert.equal((await emp.getCollateral(sponsor1)).rawValue, toWei("66.6666666666666667"));
+      let positionObject = await emp.positions(sponsor1);
+      assert.equal(positionObject.tokensOutstanding.rawValue, toWei("8"));
+
+      // Sponsor2 should have some collateral and tokens left in their position from the liquidation.
+      assert.equal((await emp.getCollateral(sponsor2)).rawValue, toWei("62.5"));
+      positionObject = await emp.positions(sponsor2);
+      assert.equal(positionObject.tokensOutstanding.rawValue, toWei("5"));
+
+      // Sponsor3 should have its full position remaining.
+      assert.equal((await emp.getCollateral(sponsor3)).rawValue, toWei("100"));
+      positionObject = await emp.positions(sponsor3);
+      assert.equal(positionObject.tokensOutstanding.rawValue, toWei("5"));
+    });
+
+    it("bot does not have enough funds to liquidate all positions", async function() {
```

@mrice32 the bot will fail silently when it doesn't have enough funds

---

Does it emit an `error` level log? why is this silent?

---

@nicholaspai I don't think we currently fail silently in that case. [This](https://github.com/UMAprotocol/protocol/blob/a7855f6a6333d60afb68a165090e6fb7f8801465/liquidator/liquidator.js#L218-L235) try catch sends out an `error` log, and I think that's where it would blow up previously if it didn't have enough inventory to liquidate. I think this is probably something we want to keep. I like that we'll do a partial liquidation now, but can we also send out an `error` log if we're forced to do that?

---

Its an `error` log so its not "silent". I meant to say its silent because the bot doesn't crash, i.e. we intentionally catch this situation.

---

There's actually an important assertion in this test that makes sure the log is an `error` log: `      assert.equal(lastSpyLogLevel(spy), "error");`